statsmodels/formula/_manager.py

```diff
@@ -0,0 +1,452 @@
+        if _engine not in ("patsy", "formulaic"):
+            raise ValueError(
+                f"Unknown engine: {_engine}. Only patsy and formulaic are supported."
+            )
+        # Ensure selected engine is available
+        msg_base = " is not available. Please install patsy."
+        if _engine == "patsy" and not HAVE_PATSY:
+            raise ImportError(f"patsy {msg_base}.")
+        elif not HAVE_FORMULAIC:
+            raise ImportError(f"formulaic {msg_base}.")
+
+        return _engine
+
+    @property
+    def engine(self):
+        """Get the formula engine."""
+        return self._engine
+
+    @property
+    def spec(self):
+        """
+        Get the model specification. Only available after calling get_arrays.
+        """
+        return self._spec
+
+    def get_arrays(
+        self,
+        formula,
+        data,
+        eval_env=0,
+        pandas=True,
+        na_action=None,
+    ) -> (
+        np.ndarray
+        | tuple[np.ndarray, np.ndarray]
+        | pd.DataFrame
+        | tuple[pd.DataFrame, pd.DataFrame]
+    ):
+        """
+        Get the design matrix and response vector from the formula and data.
+
+        Parameters
+        ----------
+        formula
+        data
+        eval_env
+        pandas
+        na_action
+
+        Returns
+        -------
+
+        """
+        if isinstance(eval_env, (int, np.integer)):
+            eval_env = int(eval_env) + 1
+        if self._engine == "patsy":
+            return_type = "dataframe" if pandas else "matrix"
+            kwargs = {}
+            if na_action:
+                kwargs["NA_action"] = na_action
+            if (
+                isinstance(
+                    formula, (patsy.design_info.DesignInfo, patsy.desc.ModelDesc)
+                )
+                or "~" not in formula
+                or formula.strip().startswith("~")
+            ):
+                output = patsy.dmatrix(
+                    formula, data, eval_env=eval_env, return_type=return_type, **kwargs
+                )
+            else:  # "~" in formula:
+                output = patsy.dmatrices(
+                    formula, data, eval_env=eval_env, return_type=return_type, **kwargs
+                )
+            if isinstance(output, tuple):
+                self._spec = output[1].design_info
+            else:
+                self._spec = output.design_info
+            return output
+
+        else:  # self._engine == "formulaic":
+            import statsmodels.formula
+
+            kwargs = {}
+            if pandas:
+                kwargs["output"] = "pandas"
+            if na_action:
+                kwargs["na_action"] = na_action
+
+            _ordering = statsmodels.formula.options.ordering
+            _formula = formulaic.formula.Formula(formula, _ordering=_ordering)
+            output = formulaic.model_matrix(_formula, data, context=eval_env, **kwargs)
+            if isinstance(output, formulaic.ModelMatrices):
```

Comment:
Does Model matrices support len? Might be good to raise for len >2 if it did.

Reply:
Yes, it does support `len`; but there are many ways that it can have length two that might be surprising. For example: `x | y` would also have length two.

Reply:
@bashtage https://github.com/matthewwardrop/formulaic/pull/207